lib/Sema/ResilienceDiagnostics.cpp

```diff
@@ -129,9 +129,12 @@ bool TypeChecker::diagnoseInlinableDeclRef(SourceLoc loc,
   DowngradeToWarning downgradeToWarning = DowngradeToWarning::No;
 
   // Swift 4.2 did not perform any checks for type aliases.
-  if (!Context.isSwiftVersionAtLeast(5) &&
-      isa<TypeAliasDecl>(D))
-    downgradeToWarning = DowngradeToWarning::Yes;
+  if (isa<TypeAliasDecl>(D)) {
+    if (!Context.isSwiftVersionAtLeast(4, 2))
+      return false;
```

Comment:
Hm. :-/ What happens in Swift 4.0 mode when Swift 5 come out?

Reply:
I’m not sure I follow?

Reply:
I thought the reason we were adding these warnings is because they could lead to linker errors in Swift 5. Maybe I'm misremembering.

Reply:
They can lead to linker errors in any release. Here is a test case:

```
struct Internal {}
typealias Alias = Internal
@inlinable func f() {
  print(Alias.self)
}
```

In 4.1, this produced no warning and failed to link when another module inlined f(). In Swift 5 mode this is an error. In 4.2 it's just a warning.

The reason it's just a warning is because there are valid code patterns you can write too, eg the underlying type of the Alias is `@usableFromInline`:

```
@usableFromInline struct Internal {}
typealias Alias = Internal
@inlinable func f() {
  print(Alias.self)
}
```

The correct way to write the above is to write, the following, which is the only variant that compiles without errors in Swift 5 mode:

```
@usableFromInline struct Internal {}
@usableFromInline typealias Alias = Internal
@inlinable func f() {
  print(Alias.self)
}
```

However, in Swift 4.1, `@usableFromInline` was not _allowed_ on type aliases at all. SwiftNIO wants to have the same source compile warning-free on both 4.1 and 4.2. So they're just not declaring their type aliases as `@usableFromInline` and promising to ensure the underlying type has the right visibility.

Once 4.2 is GM, I expect they will drop support for 4.1, and I will go back to emitting these warnings unconditionally. (And in Swift 5 mode it's an error as always).